Add import use statements section

dingo-d opened the pull request:
The PR adds rules about import `use` statements in the code and is the continuation of the additions of 'modern' PHP code in the WordPress PHP Coding Standards handbook based on the [make post](https://make.wordpress.org/core/2020/03/20/updating-the-coding-standards-for-modern-php/) by Juliette Reinders Folmer.

I'd hold off of a review until the above PR is merged.

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+
+Import `use` statements should be at the top of the file and follow the (optional) `namespace` declaration. They should follow a specific order based on the **type** of the import:
+
+1. `use` statements for classes, interfaces a traits
```
```suggestion
1. `use` statements for namespaces, classes, interfaces a traits
```

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+2. `use` statements for functions
+3. `use` statements for constants
+
+When using aliases, make sure the aliases follow the WordPress naming convention, and that you don't set the same name for alias as for the original import statement (this is not allowed: `use My\Project\Class_Name as Class_Name`). The following examples will showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.
```
```suggestion
Aliases can be used to prevent name collisions (two classes in different namespaces using the same class name).
When using aliases, make sure the aliases follow the WordPress naming convention and are unique.

The following examples showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.
```

I think we don't need to spell out the bit about using the same name (but it should be flagged by a sniff at some point).

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+
+When using aliases, make sure the aliases follow the WordPress naming convention, and that you don't set the same name for alias as for the original import statement (this is not allowed: `use My\Project\Class_Name as Class_Name`). The following examples will showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.
+
+Valid code example:
```
```suggestion
Correct:
```

jrfnl commented on wordpress-coding-standards/php.md:
````diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+
+Valid code example:
+
+```php
+<?php
+/**
+ * File doc block.
+ */
+
+namespace WordPress\subnamespace;
````
```suggestion
```php
namespace WordPress\subnamespace;
```

jrfnl commented on wordpress-coding-standards/php.md:
````diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+
+Invalid code example:
+
+```php
+<?php
+/**
+ * File doc block.
+ */
+ 
+namespace WordPress\subnamespace;
````
```suggestion
```php
namespace WordPress\subnamespace;
```

jrfnl commented on wordpress-coding-standards/php.md:
````diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+// Rest of the code
+```
+
+Invalid code example:
````
```suggestion
Incorrect:
```

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+ 
+namespace WordPress\subnamespace;
+
+use Some\Namespace \ /*comment*/ Class_A; // Comment inside an import statement.
```
I think this one can be left out again (parse error in PHP 8.0+ anyway).

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+use Some\Namespace \ /*comment*/ Class_A; // Comment inside an import statement.
+use   const   Some\Namespace\CONSTANT_A; // Extra whitespace after the use statement.
+use function Some\Namespace\function_a; // Function import after constant import. 
+use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias has wrong naming.
```
```suggestion
use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias doesn't comply with naming conventions.
```

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -135,6 +135,143 @@ do_action( "{$new_status}_{$post->post_type}", $post->ID, $post );
+use   const   Some\Namespace\CONSTANT_A; // Extra whitespace after the use statement.
+use function Some\Namespace\function_a; // Function import after constant import. 
+use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias has wrong naming.
+use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the alias.
```
```suggestion
use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, no spaces on the inside of the braces.
```

jrfnl reviewed (commented):
Just some preliminary review notes...

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
 
+### Using import `use` statements
+
+Using import `use` statements allows you to refer to an external constants, functions, classes, interfaces, and namespaces.
```
I'm struggling a bit with this sentence, I think mostly because of the use of "external" (external to what ? the answer is "the current namespace", but this will be unclear to people who are not (yet) used to working with namespaces).

Maybe we should use the definition used in the PHP manual, and link to the manual page ?

> The ability to refer to an external fully qualified name with an alias ...

Ref: https://www.php.net/manual/en/language.namespaces.importing.php

Oh and enums and traits should probably also be mentioned (if we want to be complete - I've send a [PR to the PHP docs](https://github.com/php/doc-en/pull/1748) to update the manual page for that too).

dingo-d replied:
Does this sound better?

> One of the important features of namespaces is the ability to refer to an external fully qualified name using an alias (which is also referred to as importing).
>
> PHP can alias(/import) constants, functions, classes, interfaces, traits, enums and namespaces. For more information about importing check the [PHP manual](https://www.php.net/manual/en/language.namespaces.importing.php).

GaryJones replied:
> Using import `use` statements allows you to refer to constants, functions, classes, interfaces, and namespaces that live outside of the current namespace.

?

dingo-d replied:
@GaryJones Does this mean that you agree with my proposal above, or should I just remove the alias from the paragraphs and replace it with importing?

GaryJones replied:
It means I think my suggestion should be considered as a replacement for what is currently there 😄 

```suggestion
Using import `use` statements allows you to refer to constants, functions, classes, interfaces, and namespaces that live outside of the current namespace.
```

dingo-d replied:
I'll add enums and traits to the list and your suggestion, thanks! 🙂

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+
+Import `use` statements should be at the top of the file and follow the (optional) `namespace` declaration. They should follow a specific order based on the **type** of the import:
+
+1. `use` statements for namespaces, classes, interfaces a traits
```
```suggestion
1. `use` statements for namespaces, classes, interfaces, traits and enums
```

jrfnl commented on wordpress-coding-standards/php.md:
````diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+```php
+namespace WordPress\subnamespace;
+
+use   const   Some\Namespace\CONSTANT_A; // Extra whitespace after the use statement.
````
```suggestion
use   const   Some\Namespace\CONSTANT_A; // Superfluous whitespace after the "use" and the "const" keywords.
```

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+
+The following examples showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.
+
+Correct:
```
Regarding all code samples - I realize these are copied from the Make post, but.... the examples are fine as theoretical examples, but not as real code examples as `namespace` is a reserved keyword and couldn't be used within a namespace name prior to PHP 8.0: https://3v4l.org/CNaOi

Replacing `Some\Namespace\` with `Some\Name` should be fine -  or possibly better `Prefix\SubPackage\` ?

dingo-d replied:
Totally didn't think about that 🙈 My bad. I'll replace them with some better naming 👍🏼

jrfnl commented on wordpress-coding-standards/php.md:
````diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+Correct:
+
+```php
+namespace WordPress\subnamespace;
````
```suggestion
namespace WordPress\Subnamespace;
```

Let's comply with the namespace naming rules we added in the previous PR.... (also for the incorrect example).

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+use   const   Some\Namespace\CONSTANT_A; // Extra whitespace after the use statement.
+use function Some\Namespace\function_a; // Function import after constant import. 
+use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias doesn't comply with naming conventions.
+use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, no spaces on the inside of the braces.
```
```suggestion
use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, incorrect (too little) whitespace use inside the braces.
```

Just re-read the Make post - we explicitly said that group use statements should have the "grouped" parts on new lines, so the "no spaces on the inside of the braces" needs to be rephrased. Not sure if my current suggestion covers it well enough, but probably a step in the right direction.

dingo-d replied:
Maybe:

> Extra spaces around the "as" keyword, incorrect whitespace use inside the brace opener and closer.

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+use function Some\Namespace\function_a; // Function import after constant import. 
+use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias doesn't comply with naming conventions.
+use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, no spaces on the inside of the braces.
+use Vendor\Package\{ function function_a, function function_b,
+     Class_C,
+        const CONSTANT_NAME}; // Combining different types of imports in one use statement.
```
```suggestion
use Vendor\Package\{ function function_a, function function_b,
     Class_C,
        const CONSTANT_NAME}; // Combining different types of imports in one use statement, incorrect whitespace use within group use statement.
```

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+    // Code.
+}
+
+use const \Some\Namespace\CONSTANT_D as Aliased_constant; // Import after class definition, leading backslash, naming violation.
+use function Some\Namespace\function_b as Aliased_Function; // Import after class definition, naming violation.
```
"naming violation" => "naming conventions violation" (x2)

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+[/alert]
+
+[info]
+Note that, unless you have some sort of [autoloading](https://www.php.net/manual/en/language.oop5.autoload.php) in place, the `use` statement won't automatically load whatever is being imported. You'll either need to set up autoloading, or manually load the file containing the class/function/constant using a `require/import` statement.
```
```suggestion
Note that, unless you have implemented [autoloading](https://www.php.net/manual/en/language.oop5.autoload.php), the `use` statement won't automatically load whatever is being imported. You'll either need to set up autoloading, or manually load the file containing the class/function/constant using a `require/import` statement.
```

Note: technically this is still incorrect - `use` statements don't trigger autoloading. Autoloading will only kick in when the actual class/function etc is being _used_. The `use` statements are used by PHP for the name resolution, after which the autoloading may kick in.

```php

use Some\ClassName;

class Foo {
    public function bar() {
        $a = new ClassName(); // <- This is where the autoloading takes place and only if the Foo::bar() function is actually called during the request.
    }
}
```

dingo-d replied:
> Note that, unless you have implemented [autoloading](https://www.php.net/manual/en/language.oop5.autoload.php), the `use` statement won't automatically load whatever is being imported. You'll either need to set up autoloading or load the file containing the class/function/constant using a `require/import` statement, for the aliased constructs to be loaded when used.

Better (note the addition to be loaded when used at the end)?

jrfnl reviewed (changes requested):
The usual nitpicks ;-) Thanks for working on this @dingo-d !

GaryJones commented on wordpress-coding-standards/php.md:
````diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+Incorrect:
+
+```php
+namespace WordPress\subnamespace;
````
Is this namespace line purposely incorrect (capitalization)? Either it needs a comment to say so, or it should be corrected, since the focus in this block should be on the incorrectness of the `use` statements.

dingo-d replied:
I will replace these namespaces, especially since we said that people shouldn't use `WordPress` or `wp` as a namespace.

GaryJones commented on wordpress-coding-standards/php.md:
```diff
@@ -407,6 +407,81 @@ Namespacing has no effect on variables, constants declared with `define()` or no
+While import `use` statements can already be used in WordPress Core, it is, for the moment, **strongly discouraged**.
+
+Import `use` statements are most useful when combined with namespaces and a class autoloading implementation.
+As neither of these are currently in place for WordPress Core and discussions about this are ongoing, holding off on adding import `use` statements to WordPress Core is the sane choice for now.
```
```suggestion
As neither of these are currently in place for WordPress Core and discussions about this are ongoing, holding off on adding import `use` statements to WordPress Core is the sensible choice for now.
```

jrfnl replied:
Oh, good one! 💯

dingo-d commented:
Changed the code examples to comply with the naming conventions, and removed the usage of `WordPress` as a namespace name (to avoid confusion).

Once the PR is ready the commit should be squashed and merged to keep it cleaner  🙂

GaryJones reviewed (approved):
[/]
